// @implydata/im-auth/build/client/login-entry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("es6-shim/es6-shim");
require("es7-shim");
require('./entry.css');
var React = require("react");
var ReactDOM = require("react-dom");
var js_base64_1 = require("js-base64");
var ajax_1 = require("@implydata/ajax");
var login_view_1 = require("./login-view/login-view");
var beltful_1 = require("@implydata/beltful");
ajax_1.Ajax.addErrorMonitor();
var configStr = window['__CONFIG__'];
if (!configStr)
    throw new Error('config not found');
var config = JSON.parse(js_base64_1.Base64.decode(configStr));
beltful_1.MountPoint.setRoot(config.mountPoint);
ajax_1.Ajax.mountPoint = config.mountPoint ? '/' + config.mountPoint : '';
var cont = document.getElementsByClassName('app-container')[0];
var userName = config.userName, inviteToken = config.inviteToken, accountId = config.accountId, appName = config.appName, userNameLabel = config.userNameLabel, showTerms = config.showTerms, mainDomain = config.mainDomain, mode = config.mode;
if (cont) {
    ReactDOM.render(React.createElement(login_view_1.LoginView, {
        appName: appName,
        userNameLabel: userNameLabel,
        showTerms: showTerms,
        userName: userName,
        inviteToken: inviteToken,
        accountId: accountId,
        mainDomain: mainDomain,
        mode: mode
    }), cont);
}
